File: RootTools2/src/main/java/com/rarnu/tools/neo/utils/PackageParserUtils.java
package com.rarnu.tools.neo.utils;

import android.content.ComponentName;
import android.content.IntentFilter;
import android.content.pm.*;
import android.os.Bundle;

import java.io.File;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;

@SuppressWarnings("unused")
public class PackageParserUtils {

    public final static int PARSE_IS_SYSTEM = 1 << 0;
    public final static int PARSE_CHATTY = 1 << 1;
    public final static int PARSE_MUST_BE_APK = 1 << 2;
    public final static int PARSE_IGNORE_PROCESSES = 1 << 3;
    public final static int PARSE_FORWARD_LOCK = 1 << 4;
    public final static int PARSE_EXTERNAL_STORAGE = 1 << 5;
    public final static int PARSE_IS_SYSTEM_DIR = 1 << 6;
    public final static int PARSE_IS_PRIVILEGED = 1 << 7;
    public final static int PARSE_COLLECT_CERTIFICATES = 1 << 8;
    public final static int PARSE_TRUSTED_OVERLAY = 1 << 9;

    private Object mPackageParser;

    public PackageParserUtils() {
        try {
            Class<?> cPackageParser = Class.forName("android.content.pm.PackageParser");
            Constructor<?> constructor = cPackageParser.getDeclaredConstructor();
            constructor.setAccessible(true);
            mPackageParser = constructor.newInstance();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * @param filePath
     * @param flag
     * @return PackageParser.Package
     */
    public Object parsePackage(String filePath, int flag) {
        File sourceFile = new File(filePath);
        Object mReturn = null;
        try {
            Method mParse = mPackageParser.getClass().getDeclaredMethod("parsePackage", File.class, int.class);
            mParse.setAccessible(true);
            mReturn = mParse.invoke(mPackageParser, sourceFile, flag);
        } catch (Exception e) {
        }
        return mReturn;
    }

    /**
     * @param mPackage PackageParser.Package
     * @return ArrayList<Permission>
     */
    public static ArrayList<?> packagePermissions(Object mPackage) {
        ArrayList<?> mReturn = null;
        try {
            Field fPermissions = mPackage.getClass().getDeclaredField("permissions");
            fPermissions.setAccessible(true);
            mReturn = (ArrayList<?>) fPermissions.get(mPackage);
        } catch (Exception e) {

        }
        return mReturn;
    }

    /**
     * @param mPackage PackageParser.Package
     * @return ArrayList<PermissionGroup>
     */
    public static ArrayList<Object> packagePermissionGroups(Object mPackage) {
        ArrayList<Object> mReturn = null;
        try {
            Field fPermissionGroups = mPackage.getClass().getDeclaredField("permissionGroups");
            fPermissionGroups.setAccessible(true);
            mReturn = (ArrayList<Object>) fPermissionGroups.get(mPackage);
        } catch (Exception e) {

        }
        return mReturn;
    }

    /**
     * @param mPackage PackageParser.Package
     * @return ArrayList<Activity>
     */
    public static ArrayList<Object> packageActivities(Object mPackage) {
        ArrayList<Object> mReturn = null;
        try {
            Field fActivities = mPackage.getClass().getDeclaredField("activities");
            fActivities.setAccessible(true);
            mReturn = (ArrayList<Object>) fActivities.get(mPackage);
        } catch (Exception e) {

        }
        return mReturn;
    }

    /**
     * @param mPackage PackageParser.Package
     * @return ArrayList<Activity>
     */
    public static ArrayList<Object> packageReceivers(Object mPackage) {
        ArrayList<Object> mReturn = null;
        try {
            Field fReceivers = mPackage.getClass().getDeclaredField("receivers");
            fReceivers.setAccessible(true);
            mReturn = (ArrayList<Object>) fReceivers.get(mPackage);
        } catch (Exception e) {

        }
        return mReturn;
    }

    /**
     * @param mPackage PackageParser.Package
     * @return ArrayList<Provider>
     */
    public static ArrayList<Object> packageProviders(Object mPackage) {
        ArrayList<Object> mReturn = null;
        try {
            Field fProviders = mPackage.getClass().getDeclaredField("providers");
            fProviders.setAccessible(true);
            mReturn = (ArrayList<Object>) fProviders.get(mPackage);
        } catch (Exception e) {

        }
        return mReturn;
    }

    /**
     * @param mPackage PackageParser.Package
     * @return ArrayList<Service>
     */
    public static ArrayList<Object> packageServices(Object mPackage) {
        ArrayList<Object> mReturn = null;
        try {
            Field fServices = mPackage.getClass().getDeclaredField("services");
            fServices.setAccessible(true);
            mReturn = (ArrayList<Object>) fServices.get(mPackage);
        } catch (Exception e) {

        }
        return mReturn;
    }

    /**
     * @param mPackage PackageParser.Package
     * @return ArrayList<Instrumentation>
     */
    public static ArrayList<Object> packageInstrumentation(Object mPackage) {
        ArrayList<Object> mReturn = null;
        try {
            Field fInstrumentation = mPackage.getClass().getDeclaredField("instrumentation");
            fInstrumentation.setAccessible(true);
            mReturn = (ArrayList<Object>) fInstrumentation.get(mPackage);
        } catch (Exception e) {

        }
        return mReturn;
    }

    public static ApplicationInfo packageApplicationInfo(Object mPackage) {
        ApplicationInfo mReturn = null;
        try {
            Field fApplicationInfo = mPackage.getClass().getDeclaredField("applicationInfo");
            fApplicationInfo.setAccessible(true);
            mReturn = (ApplicationInfo) fApplicationInfo.get(mPackage);
        } catch (Exception e) {

        }
        return mReturn;
    }

    public void packageCollectManifestDigest(Object mPackage) {
        try {
            Method m = mPackageParser.getClass().getDeclaredMethod("collectManifestDigest", Object.class);
            m.setAccessible(true);
            m.invoke(mPackageParser, mPackage);
        } catch (Exception e) {

        }
    }

    public void packageCollectCertificates(Object mPackage, int flags) {
        try {
            Method m = mPackageParser.getClass().getDeclaredMethod("collectCertificates", Object.class, int.class);
            m.setAccessible(true);
            m.invoke(mPackageParser, mPackage, flags);
        } catch (Exception e) {

        }
    }

    public static class Permission extends Component {
        public PermissionInfo info;
        public boolean tree;
        public PermissionGroup group;

        public static Permission fromComponent(Object mComponent) {
            Permission mReturn = new Permission();
            try {
                Field fInfo = mComponent.getClass().getDeclaredField("info");
                fInfo.setAccessible(true);
                mReturn.info = (PermissionInfo) fInfo.get(mComponent);
                Field fTree = mComponent.getClass().getDeclaredField("tree");
                fTree.setAccessible(true);
                mReturn.tree = fTree.getBoolean(mComponent);
                Field fGroup = mComponent.getClass().getDeclaredField("group");
                fGroup.setAccessible(true);
                mReturn.group = (PermissionGroup) fGroup.get(mComponent);
                mReturn.fill(mComponent);
            } catch (Exception e) {

            }
            return mReturn;
        }
    }

    public static class PermissionGroup extends Component {
        public PermissionGroupInfo info;

        public static PermissionGroup fromComponent(Object mComponent) {
            PermissionGroup mReturn = new PermissionGroup();
            try {
                Field fInfo = mComponent.getClass().getDeclaredField("info");
                fInfo.setAccessible(true);
                mReturn.info = (PermissionGroupInfo) fInfo.get(mComponent);
                mReturn.fill(mComponent);
            } catch (Exception e) {

            }
            return mReturn;
        }
    }

    public static class Activity extends Component {
        public ActivityInfo info;

        public static Activity fromComponent(Object mComponent) {
            Activity mReturn = new Activity();

            try {
                Field fInfo = mComponent.getClass().getDeclaredField("info");
                fInfo.setAccessible(true);
                mReturn.info = (ActivityInfo) fInfo.get(mComponent);
                mReturn.fill(mComponent);
            } catch (Exception e) {

            }
            return mReturn;
        }
    }

    public static class Provider extends Component {
        public ProviderInfo info;
        public boolean syncable;

        public static Provider fromComponent(Object mComponent) {
            Provider mReturn = new Provider();
            try {
                Field fInfo = mComponent.getClass().getDeclaredField("info");
                fInfo.setAccessible(true);
                mReturn.info = (ProviderInfo) fInfo.get(mComponent);
                Field fSync = mComponent.getClass().getDeclaredField("syncable");
                fSync.setAccessible(true);
                mReturn.syncable = fSync.getBoolean(mComponent);
                mReturn.fill(mComponent);
            } catch (Exception e) {

            }
            return mReturn;
        }
    }

    public static class Service extends Component {
        public ServiceInfo info;

        public static Service fromComponent(Object mComponent) {
            Service mReturn = new Service();
            try {
                Field fInfo = mComponent.getClass().getDeclaredField("info");
                fInfo.setAccessible(true);
                mReturn.info = (ServiceInfo) fInfo.get(mComponent);
                mReturn.fill(mComponent);
            } catch (Exception e) {

            }
            return mReturn;
        }
    }

    public static class Instrumentation extends Component {
        public InstrumentationInfo info;

        public static Instrumentation fromComponent(Object mComponent) {
            Instrumentation mReturn = new Instrumentation();
            try {
                Field fInfo = mComponent.getClass().getDeclaredField("info");
                fInfo.setAccessible(true);
                mReturn.info = (InstrumentationInfo) fInfo.get(mComponent);
                mReturn.fill(mComponent);
            } catch (Exception e) {

            }
            return mReturn;
        }
    }

    public static class IntentInfo extends IntentFilter {

        public boolean hasDefault;
        public int labelRes;
        public CharSequence nonLocalizedLabel;
        public int icon;
        public int logo;
        public int banner;
        public int preferred;

        public static IntentInfo fromComponent(Object mComponent) {
            IntentInfo mReturn = null;
            try {
                Field fHasDefault = mComponent.getClass().getDeclaredField("hasDefault");
                fHasDefault.setAccessible(true);
                mReturn.hasDefault = fHasDefault.getBoolean(mComponent);
                Field fLabelRes = mComponent.getClass().getDeclaredField("labelRes");
                fLabelRes.setAccessible(true);
                mReturn.labelRes = fLabelRes.getInt(mComponent);
                Field fNonLocalizedLabel = mComponent.getClass().getDeclaredField("nonLocalizedLabel");
                fNonLocalizedLabel.setAccessible(true);
                mReturn.nonLocalizedLabel = (CharSequence) fNonLocalizedLabel.get(mComponent);
                Field fIcon = mComponent.getClass().getDeclaredField("icon");
                fIcon.setAccessible(true);
                mReturn.icon = fIcon.getInt(mComponent);
                Field fLogo = mComponent.getClass().getDeclaredField("logo");
                fLogo.setAccessible(true);
                mReturn.logo = fLogo.getInt(mComponent);
                Field fBanner = mComponent.getClass().getDeclaredField("banner");
                fBanner.setAccessible(true);
                mReturn.banner = fBanner.getInt(mComponent);
                Field fPreferred = mComponent.getClass().getDeclaredField("preferred");
                fPreferred.setAccessible(true);
                mReturn.preferred = fPreferred.getInt(mComponent);
            } catch (Exception e) {

            }
            return mReturn;
        }

    }

    public static class Component {
        /**
         * PackageParser.Package
         */
        public Object owner;
        public ArrayList<IntentFilter> intents;
        public String className;
        public Bundle metaData;
        private Object mInnerComponent;

        public ComponentName getComponentName() {
            ComponentName mReturn = null;
            try {
                Method mComponentName = mInnerComponent.getClass().getSuperclass().getDeclaredMethod("getComponentName");
                mComponentName.setAccessible(true);
                mReturn = (ComponentName) mComponentName.invoke(mInnerComponent);
            } catch (Exception e) {

            }
            return mReturn;
        }

        public void fill(Object mComponent) {
            mInnerComponent = mComponent;
            try {
                Field fOwner = mComponent.getClass().getSuperclass().getDeclaredField("owner");
                fOwner.setAccessible(true);
                owner = fOwner.get(mComponent);
            } catch (Exception e) {
            }
            try {
                Field fIntents = mComponent.getClass().getSuperclass().getDeclaredField("intents");
                fIntents.setAccessible(true);
                intents = (ArrayList<IntentFilter>) fIntents.get(mComponent);
            } catch (Exception e) {
            }
            try {
                Field fClassName = mComponent.getClass().getSuperclass().getDeclaredField("className");
                fClassName.setAccessible(true);
                className = (String) fClassName.get(mComponent);
            } catch (Exception e) {
            }
            try {
                Field fMetaData = mComponent.getClass().getSuperclass().getDeclaredField("metaData");
                fMetaData.setAccessible(true);
                metaData = (Bundle) fMetaData.get(mComponent);
            } catch (Exception e) {

            }
        }
    }

}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/utils/PackageParserUtils.kt
package com.rarnu.tools.neo.utils

import android.content.ComponentName
import android.content.IntentFilter
import android.content.pm.*
import android.os.Build
import android.os.Bundle
import android.util.DisplayMetrics
import android.util.Log
import java.io.File

@Suppress("UNCHECKED_CAST")
/**
 * Created by rarnu on 4/8/16.
 */
class PackageParserUtils {

    companion object {
        val PARSE_IS_SYSTEM = 1 shl 0
        val PARSE_CHATTY = 1 shl 1
        val PARSE_MUST_BE_APK = 1 shl 2
        val PARSE_IGNORE_PROCESSES = 1 shl 3
        val PARSE_FORWARD_LOCK = 1 shl 4
        val PARSE_EXTERNAL_STORAGE = 1 shl 5
        val PARSE_IS_SYSTEM_DIR = 1 shl 6
        val PARSE_IS_PRIVILEGED = 1 shl 7
        val PARSE_COLLECT_CERTIFICATES = 1 shl 8
        val PARSE_TRUSTED_OVERLAY = 1 shl 9

        /**
         * @param pkg PackageParser.Package
         * @return ArrayList<Activity>
         */
        fun packageReceivers(pkg: Any?): MutableList<Any?>? {
            var ret: MutableList<Any?>? = null
            try {
                val fReceivers = pkg?.javaClass?.getDeclaredField("receivers")
                fReceivers?.isAccessible = true
                ret = fReceivers?.get(pkg) as MutableList<Any?>?
            } catch (e: Exception) {

            }
            return ret
        }

        /**
         * @param pkg PackageParser.Package
         * @return ArrayList<Permission>
         */
        fun packagePermissions(pkg: Any?): MutableList<Any?>? {
            var ret: MutableList<Any?>? = null
            try {
                val fPermissions = pkg?.javaClass?.getDeclaredField("permissions")
                fPermissions?.isAccessible = true
                ret = fPermissions?.get(pkg) as MutableList<Any?>?
            } catch (e: Exception) {

            }
            return ret
        }

        /**
         * @param pkg PackageParser.Package
         * @return ArrayList<PermissionGroup>
         */
        fun packagePermissionGroups(pkg: Any?): MutableList<Any?>? {
            var ret: MutableList<Any?>? = null
            try {
                val fPermissionGroups = pkg?.javaClass?.getDeclaredField("permissionGroups")
                fPermissionGroups?.isAccessible = true
                ret = fPermissionGroups?.get(pkg) as MutableList<Any?>?
            } catch (e: Exception) {

            }
            return ret
        }

        /**
         * @param pkg PackageParser.Package
         * @return ArrayList<Activity>
         */
        fun packageActivities(pkg: Any?): MutableList<Any?>? {
            var ret: MutableList<Any?>? = null
            try {
                val fActivities = pkg?.javaClass?.getDeclaredField("activities")
                fActivities?.isAccessible = true
                ret = fActivities?.get(pkg) as MutableList<Any?>?
            } catch (e: Exception) {

            }
            return ret
        }


        /**
         * @param pkg PackageParser.Package
         * @return ArrayList<Provider>
         */
        fun packageProviders(pkg: Any?): MutableList<Any?>? {
            var ret: MutableList<Any?>? = null
            try {
                val fProviders = pkg?.javaClass?.getDeclaredField("providers")
                fProviders?.isAccessible = true
                ret = fProviders?.get(pkg) as MutableList<Any?>?
            } catch (e: Exception) {

            }
            return ret
        }

        /**
         * @param pkg PackageParser.Package
         * @return ArrayList<Service>
         */
        fun packageServices(pkg: Any?): MutableList<Any?>? {
            var ret: MutableList<Any?>? = null
            try {
                val fServices = pkg?.javaClass?.getDeclaredField("services")
                fServices?.isAccessible = true
                ret = fServices?.get(pkg) as MutableList<Any?>?
            } catch (e: Exception) {

            }
            return ret
        }

        /**
         * @param pkg PackageParser.Package
         * @return ArrayList<Instrumentation>
         */
        fun packageInstrumentation(pkg: Any?): MutableList<Any?>? {
            var ret: MutableList<Any?>? = null
            try {
                val fInstrumentation = pkg?.javaClass?.getDeclaredField("instrumentation")
                fInstrumentation?.isAccessible = true
                ret = fInstrumentation?.get(pkg) as MutableList<Any?>?
            } catch (e: Exception) {

            }
            return ret
        }

        /**
         * @param pkg PackageParser.Package
         */
        fun packageApplicationInfo(pkg: Any?): ApplicationInfo? {
            var ret: ApplicationInfo? = null
            try {
                val fApplicationInfo = pkg?.javaClass?.getDeclaredField("applicationInfo")
                fApplicationInfo?.isAccessible = true
                ret = fApplicationInfo?.get(pkg) as ApplicationInfo?
            } catch (e: Exception) {

            }
            return ret
        }

    }

    private var _packageParser: Any? = null

    constructor() {
        try {
            val cPackageParser = Class.forName("android.content.pm.PackageParser")
            val constructor = cPackageParser.getDeclaredConstructor()
            constructor.isAccessible = true
            _packageParser = constructor.newInstance()

        } catch (e: Exception) {
            Log.e("LOG", "PackageParserUtils = > ${e.message}")
        }
    }

    /**
     * @return PackageParser.Package
     */
    fun parsePackage(filePath: String?, flag: Int): Any? {
        val sourceFile = File(filePath)
        var ret: Any? = null
        try {
            val parse = _packageParser?.javaClass?.getDeclaredMethod("parsePackage", File::class.java, Integer.TYPE)
            parse?.isAccessible = true
            ret = parse?.invoke(_packageParser, sourceFile, flag)
        } catch (e: Exception) {

        }
        return ret
    }

    fun packageCollectManifestDigest(pkg: Any?) {
        try {
            val m = _packageParser?.javaClass?.getDeclaredMethod("collectManifestDigest", Any::class.java)
            m?.isAccessible = true
            m?.invoke(_packageParser, pkg)
        } catch (e: Exception) {
        }
    }

    fun packageCollectCertificates(pkg: Any?, flags: Int) {
        try {
            val m = _packageParser?.javaClass?.getDeclaredMethod("collectCertificates", Any::class.java, Integer.TYPE)
            m?.isAccessible = true
            m?.invoke(_packageParser, pkg, flags)
        } catch (e: Exception) {

        }
    }

    class Permission : Component() {
        var info: PermissionInfo? = null
        var tree = false
        var group: PermissionGroup? = null

        companion object {
            fun fromComponent(component: Any?): Permission? {
                var ret = Permission()
                try {
                    val fInfo = component?.javaClass?.getDeclaredField("info")
                    fInfo?.isAccessible = true
                    ret.info = fInfo?.get(component) as PermissionInfo?
                    val fTree = component?.javaClass?.getDeclaredField("tree")
                    fTree?.isAccessible = true
                    ret.tree = fTree!!.getBoolean(component)
                    val fGroup = component?.javaClass?.getDeclaredField("group")
                    fGroup?.isAccessible = true
                    ret.group = fGroup?.get(component) as PermissionGroup?
                    ret.fill(component)
                } catch (e: Exception) {

                }
                return ret
            }
        }
    }

    class PermissionGroup : Component() {
        var info: PermissionGroupInfo? = null

        companion object {
            fun fromComponent(component: Any?): PermissionGroup? {
                var ret = PermissionGroup()
                try {
                    val fInfo = component?.javaClass?.getDeclaredField("info")
                    fInfo?.isAccessible = true
                    ret.info = fInfo?.get(component) as PermissionGroupInfo?
                    ret.fill(component)
                } catch (e: Exception) {

                }
                return ret
            }
        }
    }

    class Activity : Component() {
        var info: ActivityInfo? = null

        companion object {
            fun fromComponent(component: Any?): Activity? {
                var ret = Activity()
                try {
                    val fInfo = component?.javaClass?.getDeclaredField("info")
                    fInfo?.isAccessible = true
                    ret.info = fInfo?.get(component) as ActivityInfo?
                    ret.fill(component)
                } catch (e: Exception) {
                }
                return ret
            }
        }
    }

    class Provider : Component() {
        var info: ProviderInfo? = null
        var syncable = false

        companion object {
            fun fromComponent(component: Any?): Provider? {
                var ret = Provider()
                try {
                    val fInfo = component?.javaClass?.getDeclaredField("info")
                    fInfo?.isAccessible = true
                    ret.info = fInfo?.get(component) as ProviderInfo?
                    val fSync = component?.javaClass?.getDeclaredField("syncable")
                    fSync?.isAccessible = true
                    ret.syncable = fSync!!.getBoolean(component)
                    ret.fill(component)
                } catch (e: Exception) {

                }
                return ret
            }
        }
    }

    class Service : Component() {
        var info: ServiceInfo? = null

        companion object {
            fun fromComponent(component: Any?): Service? {
                var ret = Service()
                try {
                    val fInfo = component?.javaClass?.getDeclaredField("info")
                    fInfo?.isAccessible = true
                    ret.info = fInfo?.get(component) as ServiceInfo?
                    ret.fill(component)
                } catch (e: Exception) {
                }
                return ret
            }
        }

    }

    class Instrumentation : Component() {
        var info: InstrumentationInfo? = null

        companion object {
            fun fromComponent(component: Any?): Instrumentation? {
                var ret = Instrumentation()
                try {
                    val fInfo = component?.javaClass?.getDeclaredField("info")
                    fInfo?.isAccessible = true
                    ret.info = fInfo?.get(component) as InstrumentationInfo?
                    ret.fill(component)
                } catch (e: Exception) {
                }
                return ret
            }
        }
    }

    class IntentInfo : IntentFilter() {
        var hasDefault = false
        var labelRes = 0
        var nonLocalizedLabel: CharSequence? = null
        var icon = 0
        var logo = 0
        var banner = 0
        var preferred = 0

        companion object {
            fun fromComponent(component: Any?): IntentInfo? {
                var ret = IntentInfo()
                try {
                    val fHasDefault = component?.javaClass?.getDeclaredField("hasDefault")
                    fHasDefault?.isAccessible = true
                    ret.hasDefault = fHasDefault!!.getBoolean(component)
                    val fLabelRes = component?.javaClass?.getDeclaredField("labelRes")
                    fLabelRes?.isAccessible = true
                    ret.labelRes = fLabelRes!!.getInt(component)
                    val fNonLocalizedLabel = component?.javaClass?.getDeclaredField("nonLocalizedLabel")
                    fNonLocalizedLabel?.isAccessible = true
                    ret.nonLocalizedLabel = fNonLocalizedLabel?.get(component) as CharSequence?
                    val fIcon = component?.javaClass?.getDeclaredField("icon")
                    fIcon?.isAccessible = true
                    ret.icon = fIcon!!.getInt(component)
                    val fLogo = component?.javaClass?.getDeclaredField("logo")
                    fLogo?.isAccessible = true
                    ret.logo = fLogo!!.getInt(component)
                    val fBanner = component?.javaClass?.getDeclaredField("banner")
                    fBanner?.isAccessible = true
                    ret.banner = fBanner!!.getInt(component)
                    val fPreferred = component?.javaClass?.getDeclaredField("preferred")
                    fPreferred?.isAccessible = true
                    ret.preferred = fPreferred!!.getInt(component)
                } catch (e: Exception) {

                }
                return ret
            }
        }


    }

    open class Component {
        /**
         * PackageParser.Package
         */
        var owner: Any? = null
        var intents: MutableList<IntentFilter>? = null
        var className: String? = null
        var metaData: Bundle? = null
        private var _innerComponent: Any? = null

        constructor()

        fun getComponentName(): ComponentName? {
            var ret: ComponentName? = null
            try {
                val componentName = _innerComponent?.javaClass?.superclass?.getDeclaredMethod("getComponentName")
                componentName?.isAccessible = true
                ret = componentName?.invoke(_innerComponent) as ComponentName?
            } catch (e: Exception) {
            }
            return ret
        }

        fun fill(component: Any?) {
            _innerComponent = component
            try {
                val fOwner = component?.javaClass?.superclass?.getDeclaredField("owner")
                fOwner?.isAccessible = true
                owner = fOwner?.get(component)
            } catch (e: Exception) {
            }
            try {
                val fIntents = component?.javaClass?.superclass?.getDeclaredField("intents")
                fIntents?.isAccessible = true
                intents = fIntents?.get(component) as MutableList<IntentFilter>?
            } catch (e: Exception) {
            }
            try {
                val fClassName = component?.javaClass?.superclass?.getDeclaredField("className")
                fClassName?.isAccessible = true
                className = fClassName?.get(component) as String?
            } catch (e: Exception) {
            }
            try {
                val fMetaData = component?.javaClass?.superclass?.getDeclaredField("metaData")
                fMetaData?.isAccessible = true
                metaData = fMetaData?.get(component) as Bundle?
            } catch (e: Exception) {

            }
        }
    }


}